src/snowflake/connector/pandas_tools.py

```diff
@@ -279,20 +276,28 @@ def drop_object(name: str, object_type: str) -> None:
         )
         logger.debug(f"auto creating table with '{create_table_sql}'")
         cursor.execute(create_table_sql, _is_internal=True)
+        # need explicit casting when the underlying table schema is inferred
+        parquet_columns = "$1:" + ",$1:".join(
+            f"{quote}{snowflake_col}{quote}::{column_type_mapping[col]}"
+            for snowflake_col, col in zip(snowflake_column_names, df.columns)
+        )
     else:
         target_table_location = build_location_helper(
             database=database,
             schema=schema,
             name=table_name,
             quote_identifiers=quote_identifiers,
         )
+        parquet_columns = "$1:" + ",$1:".join(
+            f"{quote}{snowflake_col}{quote}" for snowflake_col in snowflake_column_names
+        )
 
     try:
         copy_into_sql = (
             f"COPY INTO {target_table_location} /* Python:snowflake.connector.pandas_tools.write_pandas() */ "
             f"({columns}) "
             f"FROM (SELECT {parquet_columns} FROM @{stage_location}) "
-            f"FILE_FORMAT=(TYPE=PARQUET COMPRESSION={compression_map[compression]}) "
+            f"FILE_FORMAT=(TYPE=PARQUET COMPRESSION={compression_map[compression]}{' BINARY_AS_TEXT=FALSE' if auto_create_table or overwrite else ''}) "
```

Comment:
Curious: Why do we only disable `BINARY_AS_TEXT` when creating/overwriting a table?

Reply:
because by default it is true, and we might not encounter this issue when reading into an existing table (the column is not binary). So to be safer, only when we creating a new table, we set it to false. See https://snowflake.slack.com/archives/C0UCB2YE8/p1681320980394149?thread_ts=1681257939.963869&cid=C0UCB2YE8